prov/sockets/src/sock_progress.c

```diff
@@ -2333,29 +2332,27 @@ void sock_pe_remove_rx_ctx(struct sock_rx_ctx *rx_ctx)
 static int sock_pe_progress_rx_ep(struct sock_pe *pe, struct sock_ep_attr *ep_attr,
 					struct sock_rx_ctx *rx_ctx)
 {
-	int ret = 0, i, fd, num_fds;
+	int ret = 0, i, num_fds;
 	struct sock_conn *conn;
 	struct sock_conn_map *map;
+	void *ep_contexts[SOCK_EPOLL_WAIT_EVENTS];
 
 	map = &ep_attr->cmap;
 
-        if (!map->used)
-                return 0;
+	if (!map->used)
+		return 0;
 
-        num_fds = sock_epoll_wait(&map->epoll_set, 0);
-        if (num_fds < 0 || num_fds == 0) {
-                if (num_fds < 0)
-                        SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
-                return num_fds;
-        }
+	num_fds = fi_epoll_wait(map->epoll_set, ep_contexts,
+			SOCK_EPOLL_WAIT_EVENTS, 0);
+	if (num_fds < 0 || num_fds == 0) {
+		if (num_fds < 0)
+			SOCK_LOG_ERROR("poll failed: %s\n", strerror(errno));
+		return num_fds;
+	}
 
 	fastlock_acquire(&map->lock);
 	for (i = 0; i < num_fds; i++) {
-		fd = sock_epoll_get_fd_at_index(&map->epoll_set, i);
-		if (fd == -1) /* failed to lookup fd due to connection failures */
-			continue;
-
-		conn = ofi_idm_lookup(&ep_attr->conn_idm, fd);
+		conn = ep_contexts[i];
```

Comment:
I think there's a problem with this code, which exists (and is worst) in upstream.  The issue is that the function may not progress all underlying sockets.  I believe the proposed changes help, but because SOCK_EPOLL_WAIT_EVENTS is capped at 32, if there are more sockets ready than that, they will not be progressed.  I _think_ if we add a check like:
```
if (num_fds == SOCK_EPOLL_WAIT_EVENTS)
    goto line_above_fi_epoll_wait_call;
```
right before the return statement, we can ensure that all underlying sockets that are ready are progressed.